galley/pkg/config/processing/session.go

```diff
@@ -0,0 +1,299 @@
+		s.processor.Stop()
+	}
+	scope.Debug("session.terminate: stopping sources...")
+	s.options.Source.Stop()
+
+	scope.Debug("session.terminate: signalling session termination...")
+	s.mu.Lock()
+	if s.doneCh != nil {
+		close(s.doneCh)
+		s.doneCh = nil
+	}
+	s.transitionTo(inactive)
+	s.mu.Unlock()
+}
+
+func (s *session) startProcessing() {
+	// must be called under lock.
+	if s.state != buffering {
+		panic(fmt.Sprintf("invalid state: %s (expecting buffering)", s.state))
+	}
+
+	// immediately transition to the processing state
+	o := ProcessorOptions{
+		DomainSuffix: s.options.DomainSuffix,
+		MeshConfig:   proto.Clone(s.meshCfg).(*v1alpha1.MeshConfig),
+	}
+	s.processor = s.options.ProcessorProvider(o)
+	s.buffer.Dispatch(s.processor)
+	s.processor.Start()
+	s.transitionTo(processing)
+	go s.buffer.Process()
+}
+
+func (s *session) handle(e event.Event) {
+	// Check the event kind first to avoid excessive locking.
+	if e.Kind != event.Reset {
+		s.buffer.Handle(e)
+
+		if e.Source == meshcfg.IstioMeshconfig {
+			s.handleMeshEvent(e)
+		}
+		return
+	}
+
+	// Handle the reset event
+	s.mu.Lock()
+	switch s.state {
+	case inactive, terminating:
+		// nothing to do
+
+	case starting:
+		// set the state to terminating and let the startup code complete startup steps and deal with termination.
+		s.transitionTo(terminating)
+
+	case buffering, processing:
+		s.transitionTo(terminating)
+		go s.terminate()
+
+	default:
+		scope.Warnf("Invalid session state: %v", s.state)
+	}
+	s.mu.Unlock()
+}
+
+func (s *session) handleMeshEvent(e event.Event) {
+	s.mu.Lock()
+
+	switch s.state {
+	case inactive, terminating:
+		// nothing to do
+
+	case processing:
+		scope.Infof("session.handleMeshEvent: Mesh event received during running state, restarting: %+v", e)
+		s.transitionTo(terminating)
+		go s.terminate()
+
+	case starting:
+		s.applyMeshEvent(e)
+
+	case buffering:
+		s.applyMeshEvent(e)
+		if s.meshSynced {
+			s.startProcessing()
+		}
+
+	default:
+		scope.Warnf("session.handleMeshEvent: mesh event in unsupported state '%v': %+v", s.state, e)
+	}
+
+	s.mu.Unlock()
+}
+
+func (s *session) applyMeshEvent(e event.Event) {
+	// Apply the meshconfig changes directly to the internal state.
+	switch e.Kind {
+	case event.Added, event.Updated:
+		scope.Debugf("session.handleMeshEvent: received an add/update mesh config event: %v", e)
+		s.meshCfg = proto.Clone(e.Entry.Item).(*v1alpha1.MeshConfig)
+	case event.Deleted:
+		scope.Debugf("session.handleMeshEvent: received a delete mesh config event: %v", e)
+		s.meshCfg = meshcfg.Default()
+	case event.FullSync:
+		scope.Infof("session.applyMeshEvent meshSynced: %v => %v", s.meshSynced, true)
+		s.meshSynced = true
+
+	// reset case is already handled by the time call arrives here.
+
+	default:
+		scope.Errorf("Runtime.handleMeshEvent: unrecognized event kind: %v", e.Kind)
+	}
+}
+
+func (s *session) transitionTo(st sessionState) {
+	scope.Infof("session[%d] %q => %q", s.id, s.state, st)
```

Comment:
How often does this happen? Will this create pollution in the log?

Reply:
Roughly ~4 times per meshconfig change, with a minimum of 4 per Galley lifecycle.